### PR TITLE
Json unit tests

### PR DESCRIPTION
**Pull request by vhvb1989**
# Proposal for the Internal Json test strategy

## Lite version from original full test cases

Porting the functionality tests only. Removing the build testing, fuzzing and some of the test for the features removed for the internal version like:
- _json macro
- R literal
- _json_pointer macro
- meta()
- reinterpret_bits

## Strategy

Tests would run as part of the `cpp - core` pipeline that runs nightly (or by demand on a PR running /azp run).

### CODE REVIEW

**Comment by Jinming-Hu** [comment]
When will this test case be triggered?

**Comment by vhvb1989** [comment]
> When will this test case be triggered?

Nightly. Here: https://dev.azure.com/azure-sdk/internal/_build?definitionId=1615
Or by demand. Similar to what you do when running `azp run ...`  

The regular CI builds are not affected.

**Review by Jinming-Hu — APPROVED** [review]
LGTM as long as no changes to storage code or storage CI

**Review by RickWinter — APPROVED** [review]

File: cmake-modules/download_test_data.cmake
Comment (ahsonkhan) [inline cmake-modules/download_test_data.cmake:7]: Why 2.0.0 and not 3.0.0? What version of nlohmann do these test versions correspond to? Based on your rationale, add some comments.

File: cmake-modules/download_test_data.cmake
Comment (ahsonkhan) [inline cmake-modules/download_test_data.cmake:14]: Add link to where you got this logic from. I am guessing:
https://github.com/nlohmann/json/blob/fd7a9f600712b2724463e9f7f703878ade676d6e/cmake/download_test_data.cmake#L12

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: Why was this necessary? And why are we picking 90 minutes?

File: cmake-modules/download_test_data.cmake
Comment (ahsonkhan) [inline cmake-modules/download_test_data.cmake:1]: license header missing

File: sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt:1]: license header comment missing

File: sdk/core/azure-core/test/nlohmann-json-test/src/unit-algorithms.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/src/unit-algorithms.cpp:1]: I am going to assume everything under `test/nlohmann-json-test/src/` is a **verbatim** copy of the files from https://github.com/nlohmann/json/tree/v3.8.0/test/src from the `3.8.0` git tag, and skip reviewing this. Please let me know if that's not the case or if you made any changes to it.

**Comment by ahsonkhan** [comment]
> Porting the functionality tests only. Removing the build testing, fuzzing and some of the test for the features removed for the internal version

Great idea.

File: sdk/core/azure-core/test/nlohmann-json-test/thirdparty/doctest/LICENSE.txt
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/thirdparty/doctest/LICENSE.txt:4]: I think we may need a dev-time notices and  cgmanifest update related to doctest as another unit test framework used within the SDK.

File: cmake-modules/download_test_data.cmake
Comment (vhvb1989) [reply cmake-modules/download_test_data.cmake:7]: This module is from nlohmann repo version 3.8. It points to data version 2.0.

I didn't create this file,

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: Default timeout for tests in devops is 60 min.  
Installing dependencies plus compiling the code (doing code analysis) with tests and finally running the test is taking more than one hour in the worst case for Windows.  The tests timeout before completing.  I am trying to set 90 min so we can finish

File: cmake-modules/download_test_data.cmake
Comment (vhvb1989) [reply cmake-modules/download_test_data.cmake:1]: Not ours.

File: sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt
Comment (vhvb1989) [reply sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt:1]: Not ours,  can we add our header anyway?

File: sdk/core/azure-core/test/nlohmann-json-test/src/unit-algorithms.cpp
Comment (vhvb1989) [reply sdk/core/azure-core/test/nlohmann-json-test/src/unit-algorithms.cpp:1]: The only changes are:
- replacing `#include<nlohmann/json.hpp>` for `#include<azure/core/internal/json.hpp>`
- replacing `using nlohmann::json json` for `using Azure::Core::Internal::Json json`

And some other tests were removed because it's removed/unsupported features that nlohmaan offer with global macros like `_json` .  Instead of having our own version `_az_json`, I just remove this feature. It is kind of nice to have IMO

File: cmake-modules/download_test_data.cmake
Comment (ahsonkhan) [reply cmake-modules/download_test_data.cmake:7]: Gotcha, makes sense.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: Please add that as a comment justifying making the timeout larger than the default. 90 seems like a reasonable bump from 60. Thanks for clarifying. We should keep an eye on that though, we don't want really long runs (even if it is nightly) in the worst case.

File: sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt
Comment (ahsonkhan) [reply sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt:1]: Good question, I am not sure, but it's probably fine to not add our header.

That said, in general, we should be able to modify the files as we see fit. One thing I recommend is that you add a comment at the top of the file (at least for the build related files like this cmakelist.txt and the cmake-modules file) mentioning where this is copied from, with a permalink to the file from the nlohmann repo.

The other files under `/src/` can remain as is, unchanged.

File: eng/pipelines/templates/jobs/archetype-sdk-tests.yml
Comment (vhvb1989) [reply eng/pipelines/templates/jobs/archetype-sdk-tests.yml]: Yes, It would be usually less than that thanks to vcpkg cache, but on worst case, for windows, it really takes a long time to build and install dependencies and another long time building

File: sdk/core/azure-core/test/nlohmann-json-test/src/unit-bson.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/src/unit-bson.cpp:39]: I don't think we use BSON at all (and likely won't going forward), remove this test file.

I am wondering if we should do a similar quick audit of "big items" that we know we don't need testing for.

File: sdk/core/azure-core/test/nlohmann-json-test/src/unit-msgpack.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/src/unit-msgpack.cpp:77]: Message pack is another thing we don't need tests for. It's a separate feature outside of JSON usage. Remove.

Comment (vhvb1989) [reply sdk/core/azure-core/test/nlohmann-json-test/src/unit-msgpack.cpp:77]: Since you know much more about Json, I would like to sync with you some minutes after this is merged to review the list of test files and hear from you the ones we should also remove.
I will handle that as a new PR

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Since these test files are large, I think it would be best to remove the bson and msgpack one now and avoid them in the history (we already identified them), and do a second pass later as needed (which can be separate, and lower priority).

Once those two files are gone, we can auto-merge again :)

File: CMakeLists.txt
Comment (ahsonkhan) [inline CMakeLists.txt:52]: Let's make this test disabled by default since we don't expect to be modifying the json library in daily work, improves inner dev loop.

File: sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/CMakeLists.txt:55]: remove this file

File: sdk/core/azure-core/test/nlohmann-json-test/src/unit-class_parser.cpp
Comment (ahsonkhan) [inline sdk/core/azure-core/test/nlohmann-json-test/src/unit-class_parser.cpp:1]: Since this file isn't included in the CMakeList.txt, let's remove it from the check-in/git history.

**Review by ahsonkhan — APPROVED** [review]
Delete `nlohmann-json-test/src/unit-class_parser.cpp`. Looks good otherwise.

**Pull request merged**
